src/Orleans/Utils/IKeyedServiceDescriptor.cs

```diff
@@ -0,0 +1,85 @@
+﻿
+using System;
+using System.Linq;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Orleans
+{
+    public static class KeyedServiceDescriptorExtensions
+    {
+        /// <summary>
+        /// Add supported for keyed services to a service collection
+        /// </summary>
+        public static IServiceCollection SupportKeyedServices(this IServiceCollection collection)
+        {
+            return collection.AddSingleton(typeof(IKeyedServiceCollection<,>), typeof(KeyedServiceCollection<,>));
+        }
+
+        /// <summary>
+        /// Associates a key with a specialization of a service type.
+        /// </summary>
+        public static IServiceCollection WithKey<TKey, TService, TSpecialization>(this IServiceCollection collection, TKey key)
```

Comment:
I thought the plan was to not tackle how you would register these in a generic manner (at least from our discussion, it looked like you didn't want to get there yet) but focus on how to resolve. For the registering, I'm certain that just doing type mapping will not be enough, we need to allow configuring the registered named service, which is the current prototype I'm working on.